File: file-io/xlsx/tooling/scripts/_profile-single.ts
#!/usr/bin/env npx tsx
/**
 * Profile a single XLSX file using the profiled WASM parser.
 * Usage: tsx scripts/_profile-single.ts <path-to-xlsx>
 */
import { readFileSync } from 'node:fs';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const wasmJsModule = require('../../../../compute/wasm/npm/compute_core_wasm.js');

const xlsxPath = process.argv[2];
if (!xlsxPath) {
  console.error('Usage: tsx scripts/_profile-single.ts <path-to-xlsx>');
  process.exit(1);
}

const xlsxData = readFileSync(xlsxPath);
console.log('File size:', (xlsxData.length / 1024 / 1024).toFixed(2), 'MB');

const profiled = wasmJsModule.parse_xlsx_full_profiled(new Uint8Array(xlsxData));
const timings = profiled.timings;

const ms = (us: number) => (us / 1000).toFixed(1) + 'ms';
const pct = (part: number, whole: number) =>
  whole > 0 ? ((part / whole) * 100).toFixed(1) + '%' : '-';

console.log();
console.log('=== Phase Timing Breakdown ===');
console.log('ZIP index:        ' + ms(timings.zip_index_us));
console.log(
  'Shared strings:   ' +
    ms(timings.shared_strings_us) +
    '  (' +
    pct(timings.shared_strings_us, timings.total_us) +
    ' of total)',
);
console.log(
  '  ├─ ZIP decomp:  ' +
    ms(timings.ss_zip_us) +
    '  (xml: ' +
    (timings.ss_xml_bytes / 1024 / 1024).toFixed(2) +
    ' MB)',
);
console.log('  ├─ Parse refs:  ' + ms(timings.ss_parse_refs_us));
console.log('  └─ Materialize: ' + ms(timings.ss_materialize_us));
console.log('Styles:           ' + ms(timings.styles_us));
console.log('Metadata:         ' + ms(timings.metadata_us));
console.log('Worksheets:       ' + ms(timings.worksheet_parse_us));
console.log('Serde serialize:  ' + ms(timings.serde_serialize_us));
console.log('─────────────────────────────');
console.log('Total:            ' + ms(timings.total_us));

console.log();
console.log('=== String Categories ===');
const total = timings.ss_count_total;
console.log(
  'Plain (zero-copy): ' + timings.ss_count_plain + '  (' + pct(timings.ss_count_plain, total) + ')',
);
console.log(
  'Entities-only:     ' +
    timings.ss_count_entities +
    '  (' +
    pct(timings.ss_count_entities, total) +
    ')',
);
console.log(
  'Rich text:         ' +
    timings.ss_count_rich_text +
    '  (' +
    pct(timings.ss_count_rich_text, total) +
    ')',
);
console.log('Total:             ' + total);

const result = profiled.result;
console.log();
console.log('=== Content Summary ===');
console.log('Sheets:', result.sheets.length);
console.log('Total cells:', result.stats.totalCells);
console.log('Shared strings:', result.sharedStrings.length);
console.log('Defined names:', result.definedNames.length);

console.log();
console.log('=== Per-Sheet Breakdown ===');
for (const s of result.sheets) {
  const formulas = s.cells.filter((c: any) => c.formula).length;
  console.log(
    '  ' +
      s.name.padEnd(30) +
      ' cells=' +
      String(s.cells.length).padStart(8) +
      '  formulas=' +
      String(formulas).padStart(6) +
      '  merges=' +
      s.merges.length +
      '  cf=' +
      s.conditionalFormats.length +
      '  dv=' +
      s.dataValidations.length,
  );
}

File: file-io/xlsx/parser/src/bin/profile_single.rs
//! Profile a single XLSX file with per-phase timing.
//!
//! Usage: cargo run -p xlsx-parser --bin profile_single -- <path-to-xlsx>

use std::env;
use std::fs;
use std::time::Instant;

use xlsx_parser::output::results::ParseTimings;
use xlsx_parser::pipeline::full_parse::parse_xlsx_full_native;

fn main() {
    let args: Vec<String> = env::args().collect();
    let path = args.get(1).expect("Usage: profile_single <path-to-xlsx>");

    let data = fs::read(path).expect("Failed to read file");
    println!("File: {} ({:.1} MB)", path, data.len() as f64 / 1_048_576.0);

    let mut timings = ParseTimings::zero();
    let start = Instant::now();
    let result = parse_xlsx_full_native(&data, Some(&mut timings));
    let wall = start.elapsed();

    match result {
        Ok(output) => {
            let total_cells: u32 = output.sheets.iter().map(|s| s.cells.len() as u32).sum();
            let sheets = output.sheets.len();
            println!("Sheets: {}, Cells: {}", sheets, total_cells);
            println!("Wall time: {:.1}ms\n", wall.as_secs_f64() * 1000.0);
            println!("Top-level phases:");
            println!(
                "  ZIP index:         {:>8.1}ms",
                timings.zip_index_us() / 1000.0
            );
            println!(
                "  Shared strings:    {:>8.1}ms",
                timings.shared_strings_us() / 1000.0
            );
            println!(
                "  Styles:            {:>8.1}ms",
                timings.styles_us() / 1000.0
            );
            println!(
                "  Metadata:          {:>8.1}ms",
                timings.metadata_us() / 1000.0
            );
            println!(
                "  Worksheets:        {:>8.1}ms",
                timings.worksheet_parse_us() / 1000.0
            );
            println!(
                "  Serde serialize:   {:>8.1}ms",
                timings.serde_serialize_us() / 1000.0
            );
            println!(
                "  Total reported:    {:>8.1}ms",
                timings.total_us() / 1000.0
            );
            println!("\nShared strings sub-phases:");
            println!(
                "  SS ZIP decompress: {:>8.1}ms",
                timings.ss_zip_us() / 1000.0
            );
            println!(
                "  SS parse refs:     {:>8.1}ms",
                timings.ss_parse_refs_us() / 1000.0
            );
            println!(
                "  SS materialize:    {:>8.1}ms",
                timings.ss_materialize_us() / 1000.0
            );
            println!("  SS xml bytes:      {:>8.0}", timings.ss_xml_bytes());
            println!("\nWorksheet sub-phases:");
            println!(
                "  WS decompress:     {:>8.1}ms",
                timings.ws_zip_decompress_us() / 1000.0
            );
            println!(
                "  WS cell parse:     {:>8.1}ms",
                timings.ws_cell_parse_us() / 1000.0
            );
            println!(
                "  WS cell convert:   {:>8.1}ms",
                timings.ws_cell_convert_us() / 1000.0
            );
            println!(
                "  WS postprocess:    {:>8.1}ms",
                timings.ws_postprocess_us() / 1000.0
            );
            println!(
                "  WS auxiliary:      {:>8.1}ms",
                timings.ws_auxiliary_us() / 1000.0
            );
            println!(
                "  WS aux ZIP I/O:    {:>8.1}ms",
                timings.ws_aux_zip_io_us() / 1000.0
            );
        }
        Err(e) => {
            eprintln!("Parse error: {}", e);
        }
    }
}
